### DISCUSSION TITLE
Added: LDM Notes for October 25th, 2021

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#5342] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-10-25.md

## Agenda

1. Required members
2. Delegate type argument improvements

**Comment by HaloFour** [comment]
"Delegate type argument improvements" feels like an awful lot of work to avoid declaring custom delegates to cover those cases where `Action<...>`/`Func<...>` aren't suitable.  Is there a plan to eventually extend these generic improvements across types and methods?  If not it would just feel really weird that I could define `Func<string, ref int, bool>` but not `M<T1, T2, TR>(Func<T1, T2, TR> fn)` that could accept such a generic delegate.

**Reply by 333fred** [reply]
That would be one of the things that was brought up in the notes 😛

**Reply by HaloFour** [reply]
Understood, was curious if there were more details than glossed over in the notes.  😀 

I think we're on the same page, and it would be really interesting if generics could be expanded to support `ref`/`in`/`out` parameters, but I imagine that it wouldn't take much to totally blow up the assumptions that the runtime currently allows us to make about the generic type parameters.  Especially if you try to get `void` in there too.

**Reply by RikkiGibson** [reply]
It's a tough position to be in. I'll be straight up and say that deciding what it means when a method implementation uses a type parameter that may or may not be `ref` was not a goal of my proposal. I understand why the language team wants at least an informed dive into what that would mean and how it intersects with this, though. I'm also sympathetic that there's plenty of other missing pieces in the proposal when it comes to generalizing on delegates, such as attributes or large numbers of parameters.

What I primarily want to accomplish is to move away from from the grab-bag of "Action+Func+custom delegates", to a general purpose structural function type which renders named delegate types unnecessary in general--at most you might want to use a name as an alias to some structural function type. It would also have to somehow avoid bifurcation in the language. We don't want library authors who have been using delegate types in public API to have to add a whole new set of APIs just to take advantage of a shiny new function type. It's possible that this is a step in that direction, but a lot more research is needed to decide if that is really the case or not.

Related to the effort of fully generalizing these type arguments: I have been tinkering with the concept of a `T : any` constraint which would permit a slew of restricted types as type arguments. One thing that has occurred to me fairly quickly is that allowing an `out` type argument for a parameter with such a constraint doesn't seem to make sense. I can't picture usefully generalizing over a parameter which may or may not be `out`, for example, outside of the context of simply constructing a method signature for use later.

**Reply by TahirAhmadov** [reply]
> 
> What I primarily want to accomplish is to move away from from the grab-bag of "Action+Func+custom delegates", to a general purpose structural function type which renders named delegate types unnecessary in general--at most you might want to use a name as an alias to some structural function type.

Following that logic, why not abolish classes altogether, at least those which act as POCOs/DTOs, and use `Tuple`s for everything? 

The importance of having a distinct type is not in the signature, but in the purpose of the delegate (or a POCO class). It was a mistake to use `Func<T, bool>` for LINQ `Where` methods; `Predicate<T>` would've been much more concise. It was an even bigger mistake to start using `Func<,,,,,>` type delegates. 

Making generic arguments for `Func` allow `ref` keyword, etc. isn't a problem by itself - perhaps it can be useful for generics in general. However, the motivation is completely flawed. This is time and money better spent at much more important improvements.

**Reply by CyrusNajmabadi** [reply]
Classes do more than just provide names. If that's an they did, there would actually be good reasons to wonder about that. However, classes tie together collections of state and functionality together.  

A delegate doesn't do that. 

There are currently valuable parts of a Delegate type. But there's no string reason this has to be tied to a specific type and couldn't just be part of the design of a structural delegate value.

**Reply by TahirAhmadov** [reply]
> 
> 
> Classes do more than just provide names. 

That was literally in the very first sentence of my comment - talking about POCOs specifically.

> If that's a**ll** they did, there would actually be good reasons to wonder about that. 

No there would not be. Imagine a POCO named "Company" which has a "Name" property, and a POCO named "Country" which also has a "Name" property. Would you want to abolish the classes "Company" and "Country" and just use `Tuple<string>`?

> But there's no str**o**ng reason this has to be tied to a specific type and couldn't just be part of the design of a structural delegate value.

This line of reasoning - do we have a strong reason or not - is precisely the underlying cause of many articles at thedailywtf.com. Good code is not one that is shipped because it just barely works; it's readable, maintainable, reliable, performant and secure code. The only way to deliver such high quality code (even understanding that nothing is ideal) is to design an elegant language; a "pit of success". This loose typing and convertibility between `Tuple`s and delegates is the exact opposite of that. Frankly, it has a strong smell of JavaScript (in the worst possible meaning of that).

**Reply by TahirAhmadov** [reply]
PS. To clarify, "Country" should never be convertible to "Company" (even if both are basically `Tuple<string>`), and "ParseStringToIntDelegate" should never be convertible to "GetEmailCountByCompanyNameDelegate" (even if both are `Func<string, int>`).

**Reply by CyrusNajmabadi** [reply]
> Would you want to abolish the classes "Company" and "Country" and just use 

Depending on the domain, that might be totally fine yes. I find myself rarely using pocos, so the idea that they're somehow sacrosanct doesn't hold much weight for me. 

> ; it's readable, maintainable, reliable, performant and secure code

I don't see anything about the recent discussion that negatively affects any of that. 

Objectively, it is neutral on most of that. Subjectively I find it better for some parts as well. 

You're welcome to disagree. That's the nature of subjective analysis. But I think you'll find that your views here definitely don't align. 

As an example, I see no value with Predicate<T>, And there are a issues with it (including lack of composability).

**Reply by CyrusNajmabadi** [reply]
> GetEmailCountByCompanyNameDelegate

And yet, I might have a transformational pipeline that composes arbitrary transformations. I cannot compose arbitrary delegate types this way as the transformation system now needs to know about all those types. 

Composability is a strong goal here as well, and that means having components talk about functionality *generically* (both in the abstract and literal-.NET senses).

**Reply by TahirAhmadov** [reply]
> > Would you want to abolish the classes "Company" and "Country" and just use
> 
> Depending on the domain, that might be totally fine yes. I find myself rarely using pocos, so the idea that they're somehow sacrosanct doesn't hold much weight for me.

In that case, why even use `Tuple`s? Why not make everything a dictionary of name-value pairs? Granted, it's the next step and not quite the same thing, but it's in the same direction.

> > ; it's readable, maintainable, reliable, performant and secure code
> 
> I don't see anything about the recent discussion that negatively affects any of that.

Objectivity galore.
Readable: `Func<,,,>` doesn't say anything about what it's supposed to do; named delegate does.
Maintainable: If our hypothetical "GetEmailCountByCompanyNameDelegate" needs to be changed, say, to add a parameter like `bool publicEmailsOnly`, it can be done in one place, then build will immediately tell you everywhere you need to change the target methods and callers of the delegate. Whereas with `Func<string, int>`, there is literally no reliable way of achieving this objective. Not only you would need to rely on text search and replace (because `Func` is used for everything, so find all references doesn't work), but you would need to be careful to discriminate between the "get email count" delegates and "parse string to int" delegates.
Reliable: simple typo and you are passing the wrong delegate to a method, because they're all `Func`s - boom, YSOD.
Performant: correct me if I'm wrong, but generic `Func<,,>` variants have to be constructed - meaning, there is a startup cost. Whereas a concrete delegate type is loaded with the assembly right away and available to be used. Granted even if true this is a miniscule difference, but a difference nonetheless.
Secure: this isn't directly in play, but it can negatively impacted by previously listed issues.

> As an example, I see no value with Predicate, And there are a issues with it (including lack of composability).

Can you explain what you mean by "lack of composability"?

PS. Re "maintainable" - given the potential for whitespace differences, you're talking about a regex search and replace: "Func\s*<\s\*string\s\*,\s\*int\s\*>" - exactly the kind of thing you want programmers to do on large and complex solutions - **NOT**.

**Reply by TahirAhmadov** [reply]
> And yet, I might have a transformational pipeline that composes arbitrary transformations. I cannot compose arbitrary delegate types this way as the transformation system now needs to know about all those types.
 
Can you please provide a simple example of what you mean by "transformational pipeline that composes arbitrary transformations"? Just a link would be fine.

> Composability is a strong goal here as well, and that means having components talk about functionality _generically_ (both in the abstract and literal-.NET senses).

Again, if you could suggest any pointers so that I can comprehend the meaning of this, it would be appreciated.

**Reply by En3Tho** [reply]
There is no meaning in making DelegateGetCompanyByCompanyName taking an arbitrary string. If you're willing to name every delegate you might use then you might instead just define "CompanyName" type. And Func<CompanyName, Company> is more that self-explanatory. Do not mistake "name safety" for type safety.

**Reply by YairHalberstadt** [reply]
@TahirAhmadov 

I almost never use named delegates and I just haven't been bitten by the issues you describe. Why I could reasonably imagine that unnamed generic delegates might have the issues you describe, my actual experience is that this hasn't happened.

This is as opposed to e.g. nullable reference types, where I was regularly encountering NullReferenceErrors before they appeared.

**Reply by TahirAhmadov** [reply]
> There is no meaning in making DelegateGetCompanyByCompanyName taking an arbitrary string. If you're willing to name every delegate you might use then you might instead just define "CompanyName" type. And Func<CompanyName, CompanyType> is more that self-explanatory. Do not mistake "name safety" for type safety.

Great point - I would definitely pass around my "Company" POCO, and not a `string companyName`. Perhaps I picked the wrong example to make my point; a much better difference would be "ParseStringToIntDelegate" vs. "GetStringHashCodeDelegate". Again, please understand - these are hypothetical examples, I never had a need to pass around neither of those in my production code. I'm just trying to make the point that the signature isn't the only thing that makes a delegate. And I think you agree with me; you don't like to pass `string` around, because it could be a company name, or a country name, etc. Follow that exact logic and you'll see how passing around `Func<string, int>` is equally bad.

**Reply by TahirAhmadov** [reply]
> 
> 
> @TahirAhmadov
> 
> I almost never use named delegates and I just haven't been bitten by the issues you describe. Why I could reasonably imagine that unnamed generic delegates might have the issues you describe, my actual experience is that this hasn't happened.
> 
> This is as opposed to e.g. nullable reference types, where I was regularly encountering NullReferenceErrors before they appeared.

I'll tell you something even funnier, the nullable reference types didn't actually decrease my NREs all that much; they were near zero before and are near zero now - despite me marking all my projects nullable=enabled and investing countless hours into changing all the signatures. However, I still wholeheartedly support NRTs for two reasons: 1) I myself can make a mistake later despite not being prone to this particular type of mistakes currently, and 2) personal experience isn't as important - other teams have different members with various levels of experience. 

However, I did start noticing issues with `Func<,,,>` delegates in some of my most complex projects. Just the readability of named delegate types made such a huge difference for me. And yes, I had to modify the signature of a delegate and named delegates made it much easier.

**Reply by CyrusNajmabadi** [reply]
> In that case, why even use Tuples? Why not make everything a dictionary of name-value pairs? 

Performance and strong typing.

**Reply by TahirAhmadov** [reply]
> > In that case, why even use Tuples? Why not make everything a dictionary of name-value pairs?
> 
> Performance and strong typing.

Obviously performance is affected by dictionary lookups, and again I was purposefully making a ridiculous example. However, when it comes to strong typing, tuples being convertible between one another solely because the member types match, is an erosion of strong typing.

**Reply by HaloFour** [reply]
Tuples exist for those cases where the container is meaningless but you need the means to convey more than one loosely associated value.  Similarly the `Action<...>`/`Func<...>` families of delegates exist for the same reason, where the operation is not as important as the signature.  Neither are intended to replace strong types.

**Reply by TahirAhmadov** [reply]
> Tuples exist for those cases where the container is meaningless but you need the means to convey more than one loosely associated value. Similarly the `Action<...>`/`Func<...>` families of delegates exist for the same reason, where the operation is not as important as the signature. Neither are intended to replace strong types.

I agree, my concern is that an effort is being made to enhance the `Func<>` etc. to encompass all the possible situations, which is way outside the scope of what you are referring to - things like LINQ delegates (my preference for `Predicate<T>` over `Func<T,bool>` aside).

**Reply by HaloFour** [reply]
The purpose of the project was indeed to make it possible to use `Action<...>`/`Func<...>` in more circumstances as currently the cases where you're forced to use different delegate types is driven by the limitations of generics in the CLR and not based on business or logic reasons.  I don't think this project changes what you should weigh when using a tuple or general-purpose delegate vs. crafting your own named type, but it removes an artificial technical barrier which forces the hand of the developer.

It's your choice to prefer named types over tuples and general-purpose delegates, but it seems that the trend in the language/runtime has been to move away from having to declare those types just for the sake of defining the shape of the inputs/outputs to generic code.

**Reply by TahirAhmadov** [reply]
> The purpose of the project was indeed to make it possible to use `Action<...>`/`Func<...>` in more circumstances as currently the cases where you're forced to use different delegate types is driven by the limitations of generics in the CLR and not based on business or logic reasons. 

I [listed](https://github.com/dotnet/csharplang/discussions/5342#discussioncomment-1539693) the reasons why `Func<>` should be avoided altogether in my reply to Cyrus above.

> I don't think this project changes what you should weigh when using a tuple or general-purpose delegate vs. crafting your own named type, but it removes an artificial technical barrier which forces the hand of the developer.

I understand that, I'm not afraid that named delegate types would be removed from C#. However, investing a significant amount of effort into making generic types more flexible ostensibly to allow a more broad usage of what I've shown to be anti-patterns (specifically generic delegates) misallocates limited resources.

> It's your choice to prefer named types over tuples and general-purpose delegates, but it seems that the trend in the language/runtime has been to move away from having to declare those types just for the sake of defining the shape of the inputs/outputs to generic code.

And that's exactly my point - the signature is just one part of a delegate, not all of it. Whereas `ValueTuple` serves a very real purpose - for example, a method needs to return multiple values, and this combination of the return values only makes sense as the return type of a method and doesn't deserve its own class - this scenario doesn't apply to delegates. 

I would be happy to change my mind on this - if you can, please share an example where `Func<,,,>` type delegate is preferable to a named delegate type, especially with complicated parameters like `ref`/etc.

**Reply by HaloFour** [reply]
@TahirAhmadov 

You're welcome to your opinion that the use of the general purpose delegate families `Action<...>`/`Func<...>` represents an anti-pattern that should be discouraged, but it's pretty clear that the C# language and .NET runtime teams do not share that opinion given the expansion of inference for lambdas in C# 10.0.

**Reply by TahirAhmadov** [reply]
> You're welcome to your opinion that the use of the general purpose delegate families `Action<...>`/`Func<...>` represents an anti-pattern that should be discouraged, but it's pretty clear that the C# language and .NET runtime teams do not share that opinion given the expansion of inference for lambdas in C# 10.0.

1) C# and .NET teams holding any particular opinion doesn't make it correct automatically - despite doing a great job overall; 
2) reading through that page, I'm a little confused as to what real world usages of the new syntax forms would be. For example, the inference of a lambda type to a `Func<,,,>` type with `var` keyword; I can imagine how developers are annoyed with having to type `Func<string, object, int>` or something like that and want `var` keyword; but the problem here is the original decision to allow `Func<,,,>` types in the first place. This problem wouldn't have existed if the guidance from the beginning was to always create delegate types; a simple delegate type identifier is easy to type with auto-complete (as opposed to the `Func` and angle bracket and generics and commas). This is precisely a situation where a bad decision in the past breeds more bad decisions going forward.

**Reply by HaloFour** [reply]
@TahirAhmadov 

> C# and .NET teams holding any particular opinion doesn't make it correct automatically

> This is precisely a situation where a bad decision in the past breeds more bad decisions going forward.

"Correct" and "bad" are subjective.  The opinions of the language and runtime teams inform their decisions regarding design and prioritization.  They clearly do not feel that these decisions are bad, or that further investment in this space would be bad.  Language design isn't about pleasing everyone.

**Reply by CyrusNajmabadi** [reply]
> This problem wouldn't have existed if the guidance from the beginning was to always create delegate types; a simple delegate type identifier is easy to type with auto-complete

But it would often pointlessly force me to both define such a type, and then repeat that name on many places annoyingly. 

This is the point here. There are many cases where that it strictly a tax. Indeed, I find it to be the majority case. 

This is similarly why tuples are valuable, despite it being possible to use a named type instead. I often do not need it, so it becomes ceremony for ceremony's sake. 

The restriction on Func/Action were not there because we thought they were a good thing. They were there because of runtime limitations. As such, of we can get through the limitations then this makes those types act as we've wanted them to act since they were added 15+years ago.

**Reply by RikkiGibson** [reply]
If "structural delegates" were to come about then it would still be possible to apply coding patterns which enforce some domain validity of the underlying value. e.g. people have been known to do 

```cs
struct StudentId
{
    int underlying;
}

struct BuildingId
{
    int underlying;
}
```

I don't consider this a typical requirement for the way people use function types, but the original delegate types feature more or less makes these claims about domain validity required. I think it is better if the default is to have compatible function signatures be implicitly convertible to each other with minimal allocation, and if the user has some desire to use the type system to add a guard rail for valid use of their domain objects, they would always be free to do that.

**Reply by TahirAhmadov** [reply]
> 
> 
> @TahirAhmadov
> 
> > C# and .NET teams holding any particular opinion doesn't make it correct automatically
> 
> > This is precisely a situation where a bad decision in the past breeds more bad decisions going forward.
> 
> "Correct" and "bad" are subjective. The opinions of the language and runtime teams inform their decisions regarding design and prioritization. They clearly do not feel that these decisions are bad, or that further investment in this space would be bad. Language design isn't about pleasing everyone.

By no means I want the language to please me. I'm not claiming that such and such feature is bad because I don't like it. I brought forward some objective arguments. Obviously it is up to the language team to decide if they want to listen to my feedback. However your comment is one more assertion of authority and doesn't address any of the points I made previously. If and when you decide to actually address these points, I would be happy to ingest that information.

**Reply by TahirAhmadov** [reply]
> > This problem wouldn't have existed if the guidance from the beginning was to always create delegate types; a simple delegate type identifier is easy to type with auto-complete
> 
> But it would often pointlessly force me to both define such a type, and then repeat that name on many places annoyingly.

Mind you, I'm not against generics in delegate types altogether. In my ideal world, `Predicate<T>` would be used for `Where` LINQ method; `SelectDelegate<TFrom, TTo>` for `Select`; etc. Most of the existing lambdas would continue to be interpreted just as they are now. If a programmer needs a generic delegate, like `GetOrganizationEmailCountDelegate<T> where T : Organization`, that's fine, too. Between built in delegate types, and custom ones created to fit the purpose at hand, using generics as needed, there wouldn't be some crazy number of delegate types. Also, delegate type declaration is a one-liner and easy to type; it's literally 1% of the work involved, if you factor in the callees, callers, and other stuff in a typical project.

> The restriction on Func/Action were not there because we thought they were a good thing. They were there because of runtime limitations. As such, of we can get through the limitations then this makes those types act as we've wanted them to act since they were added 15+years ago.

I'm not arguing that the restrictions are a good thing; it's entirely possible that `ref` generic types are needed. My point is, the intended reason to make `Func` even easier to use adds to a different problem - the existence of `Func` altogether. 

If `ref`/etc. generic types are needed, I would say it should be considered in the context of all generics, and the implication of such a change considered when it comes to types (including delegates), methods, reflection, runtime, etc. Analysis should be performed as to how much such an improvement is needed and the cost of doing it. As part of that analysis, IMO, the benefit for delegates like `Func` should be treated as zero, if not negative, because it further discourages named delegates.

**Reply by HaloFour** [reply]
@TahirAhmadov 

> By no means I want the language to please me. I'm not claiming that such and such feature is bad because I don't like it. I brought forward some objective arguments. Obviously it is up to the language team to decide if they want to listen to my feedback. However your comment is one more assertion of authority and doesn't address any of the points I made previously. If and when you decide to actually address these points, I would be happy to ingest that information.

And members of the team have responded to your arguments in this thread, so I don't know what more to tell you.  You weigh the pros and cons of these features differently.  I'd suggest further discussion could take place on Discord if you want but I have a feeling that our conversations are not going to convince you or vice versa.

**Reply by TahirAhmadov** [reply]
> enforce some domain validity of the underlying value. 
> I don't consider this a typical requirement for the way people use function types, but the original delegate types feature more or less makes these claims about domain validity required. I think it is better if the default is to have compatible function signatures be implicitly convertible to each other with minimal allocation, and if the user has some desire to use the type system to add a guard rail for valid use of their domain objects, they would always be free to do that.

1. Even if you create a `StudentId` type to enforce strong type instead of `int`, it still wouldn't protect you against mixing up `Func<StudentId, bool> isStudentSick` and `Func<StudentId, bool> isStudentSuspended`.
2. As I showed above, sometimes it is purposefully a built-in type like `string`.
3. Even if `Func<StudentId, bool>` is only used for `isStudentSick`, meaning you successfully used the type as a way to enforce domain validity, this merely means that the expense of creating a type has been pushed down the chain. It's like buying an electric car in a place where 100% of the electricity generated by coal (a hypothetical place). If you want to spend the effort to create a type, why not just create a delegate?
4. The biggest issue, though, is that it makes this "guard rail" optional. Given that the language made it easy to use `Func<>`, and the coder has a choice between using `int` directly, or creating a `StudentId` class, it's much more likely than not that he will just use `int`, given the real life limitations: everything is due yesterday, I want to go our for coffee instead of coding, etc.

**Reply by RikkiGibson** [reply]
> 1. Even if you create a `StudentId` type to enforce strong type instead of `int`, it still wouldn't protect you against mixing up `Func<StudentId, bool> isStudentSick` and `Func<StudentId, bool> isStudentSuspended`.

This seems equivalent in nature to accidentally passing `bool isStudentSick` as an argument for the `bool isStudentSuspended` parameter of some method. Perhaps people should also have to declare their own bool types to avoid mixing up different flags in their domain model which have incompatible meanings? The answer is of course "no" for the typical case. But again people are welcome to introduce wrappers or custom named types for such flags if they feel the type system behavior is helping them prevent misuse.

```cs
enum IsStudentSick { No, Yes }
enum IsStudentSuspended { No, Yes }
```

**Reply by CyrusNajmabadi** [reply]
> t still wouldn't protect you against mixing up Func<StudentId, bool> isStudentSick and Func<StudentId, bool> isStudentSuspended.

Correct.  the same way nothing protects you if you have:

`bool isStudentSick` and `bool isStudentSuspended`, and then you check the wrong one.   If you need to protect against that, then by all means, add guardrails in the type system.  However, not all potential accidents must be prevented and not every single thing needs to be modeled as an explicit type.  It's acceptable and reasonable for users to reuse types when they feel that such guardrails are not necessary and that teh structure that type provides is what is giving htem value.

**Reply by CyrusNajmabadi** [reply]
Lol. as rikki and i having the exact same thought/example :)

**Reply by CyrusNajmabadi** [reply]
> If you want to spend the effort to create a type, why not just create a delegate?

That's the point.  I do *not* want to spend the effort to create the type. Indeed, in many domains i work in, it is the *common* case that i don't want to create a type here.  

For thigns like tuples it can depend.  There are a ton more cases where i do want the type (often because it's either grown so large to be unweildy as a tuple, or because it does convey a combination of state and logic).  However, whereas with types, there is a lot often gained by moving to a full type, the same is much less true for delegates.  I can certainly gain in some cases, but it's often exceedingly little.  And, in some cases, it's nothing at all and the type serves just as an onerous hoop to jump through (often because of historical decisions).

**Reply by CyrusNajmabadi** [reply]
> the benefit for delegates like Func should be treated as zero, if not negative, because it further discourages named delegates.

Named delegates remain as relevant as before for anyone who feel they are appropriate for their domain.  If named delegates were needed just to workaround a limitation in another area then that's not an argument in favor of named delegates, and it's not a reason to continue to prop them up.  Named delegates should survive and be used on their own merits, and they should not be used as a cudgel to deal with very real and unpleasant limitations in other areas of the language.

**Reply by TahirAhmadov** [reply]
> This seems equivalent in nature to accidentally passing `bool isStudentSick` as an argument for the `bool isStudentSuspended` parameter of some method. 

That's true, and your enum solution makes sense in some cases. It could be "No, Yes, YesWithDoctorNote" and stuff.

Keep in mind that it's easy to argue for `Func<>` when it's just 2 generic arguments - `Func<Student, bool>` kind of thing. Obviously we cannot remove `Func<>` and people will use it going forward for these simple situations. My concern was with complex delegates, like `Action<string, string, Action<IDataReader>> executeDbQuery` type of delegate: multiple arguments, a lot of typing (similar to creating a delegate type); it's unclear whether the first argument is a connection string and second one is the query, or the other way around.

**Reply by CyrusNajmabadi** [reply]
> If ref/etc. generic types are needed, I would say it should be considered in the context of all generics

Note: we are doing that (and i thought the LDM notes made that clear).  Specifically, we are considering the greater context here.  That doesn't mean we'll end up deciding to solve the greater context (or that there is a solution).  But we are certainly cognizant of it and will have an open mind to where we want to be moving to.

If it helps, several LDM members (myself included) made it explicitly clear that we should not narrowly scope the discussion/design to just trying to solve "`ref` in `Func/Action`".  We should broadly look at hte landscape here and think where we'd like to be in the short, medium and long term.  We may still only take small steps, but we will certainly be looking to see if we can larger ones (now or in the future).

**Reply by CyrusNajmabadi** [reply]
>  it's unclear whether the first argument is a connection string and second one is the query, or the other way around.

As with basic data types, tuples, etc. etc., there's a spectrum here.  We've never advocated using things in a confusing fashion.  Rather, we give the tools and let teams decide what styles and choices are best for them.

**Reply by HaloFour** [reply]
> it's unclear whether the first argument is a connection string and second one is the query, or the other way around.

And thus was born [Apps Hungarian Notation](https://en.wikipedia.org/wiki/Hungarian_notation#Systems_Hungarian_vs._Apps_Hungarian), to denote a logical type apart from the actual data type that can be visually identified as different.

Also, with C# 10.0 `record struct` gets us pretty close to zero-cost type-safe wrappers with very little fuss.

```cs
public record struct StudentId(string id);
```

**Reply by jmarolf** [reply]
In general, @TahirAhmadov the design and use of delegates at their inception matches what you describe "You should define your own delegates that match your given domain" and I think this still makes sense in a lot of domains. However, for libraries that are not domain specific this quickly fell out of practice when generics where introduced. In the framework these are the only delegates beyond action/func that are [ever defined](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Action.cs#L29-L33)

```csharp
public delegate int Comparison<in T>(T x, T y);
public delegate TOutput Converter<in TInput, out TOutput>(TInput input);    
public delegate bool Predicate<in T>(T obj); 
```

And these delegates are defined for a good reason: predicates make it clearer the delegate you are passing is used for filtering/checking. `Comparison` and `Converter` are also special use and helpful to clarify how a given delegate will be used.

So, I would say the more general a library is the less you should need to define custom delegates. For this reason I think this change is going to be super helpful for library authors but probably less useful apps.

**Reply by RikkiGibson** [reply]
I think one thing we could do to alleviate some of the pain around usability of "custom" delegates would be to prioritize dotnet/roslyn#46190.

**Reply by TahirAhmadov** [reply]
> 
> 
> > it's unclear whether the first argument is a connection string and second one is the query, or the other way around.
> 
> As with basic data types, tuples, etc. etc., there's a spectrum here. We've never advocated using things in a confusing fashion. Rather, we give the tools and let teams decide what styles and choices are best for them.

I agree that bad developers will always find a way to shoot themselves in their feet. From the perspective of language design, it's one thing to provide "good" options, and another to provide "bad" options. Whereas with `Func<T1, T2>` it's difficult to make a mistake like what I described, with `Func<T1, T2, T3, T4>` it is much easier to do so. I just hope LDM doesn't prioritize changing the language in the direction of making it easier to use things like `Func<T1, T2, T3, T4>` - which are objectively bad for all the reasons I've listed so far.

**Reply by TahirAhmadov** [reply]
> 
> 
> In general, @TahirAhmadov the design and use of delegates at their inception matches what you describe "You should define your own delegates that match your given domain" and I think this still makes sense in a lot of domains. However, for libraries that are not domain specific this quickly fell out of practice when generics where introduced. In the framework these are the only delegates beyond action/func that are [ever defined](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Action.cs#L29-L33)
> 
> ```cs
> public delegate int Comparison<in T>(T x, T y);
> public delegate TOutput Converter<in TInput, out TOutput>(TInput input);    
> public delegate bool Predicate<in T>(T obj); 
> ```
> 
> And these delegates are defined for a good reason: predicates make it clearer the delegate you are passing is used for filtering/checking. `Comparison` and `Converter` are also special use and helpful to clarify how a given delegate will be used.
> 
> So, I would say the more general a library is the less you should need to define custom delegates. For this reason I think this change is going to be super helpful for library authors but probably less useful apps.

I agree, and again, I'm not against generic delegates in principle; the ones you listed, and the simple versions of `Action` and `Func<T>` are usually fine, because the intent is obvious - if the intent is "perform an action", then `Action` fits perfectly. I was concerned that what I read was people saying, named delegates are all but obsolete, all delegates will be `Action` and `Func` from now on, and we'll massage whatever parts of the language that are needed to make that workable.

In my work, I do end up making libraries, not as much for public consumption but to be re-used by me and multiple teams of my larger clients. And even in libraries, I used to default to `Func<T1, T2, T3, T4>` type delegates - because I thought, hey it's simpler - until I realized that the readability and maintainability suffer as a result.

**Reply by TahirAhmadov** [reply]
> 
> 
> > it's unclear whether the first argument is a connection string and second one is the query, or the other way around.
> 
> And thus was born [Apps Hungarian Notation](https://en.wikipedia.org/wiki/Hungarian_notation#Systems_Hungarian_vs._Apps_Hungarian), to denote a logical type apart from the actual data type that can be visually identified as different.
> 
> Also, with C# 10.0 `record struct` gets us pretty close to zero-cost type-safe wrappers with very little fuss.
> 
> ```cs
> public record struct StudentId(string id);
> ```

That's a good feature, and I would say - call me a strong typing nazi - I would use the above **and** named delegates. Basically, I already do `delegate void ExecuteDbDelegate(string connStr, string query);`, and in some cases, I'll probably change it to something like `delegate void ExecuteDbDelegate(ConnStr connStr, Query query);`. (Disclaimer: I don't actually have a delegate like this - just came up with this as an example.)

**Reply by HaloFour** [reply]
You're free to always make that decision, but the language team isn't going to put arbitrary restrictions on the feature based on the arity of the parameters.  This is true already with tuples and with naturally-typed lambdas both of which are capable of exceeding the arity of the .NET types to which they map by default.  The language team disagrees that this is a "bad" thing as does seemingly a good percentage of the audience who find the need to define custom delegates a nuisance.

**Reply by jmarolf** [reply]
> In my work, I do end up making libraries, not as much for public consumption but to be re-used by me and multiple teams of my larger clients. And even in libraries, I used to default to Func<T1, T2, T3, T4> type delegates - because I thought, hey it's simpler - until I realized that the readability and maintainability suffer as a result.

I do think `Func<...>` has some ergonomics issues where you need to remember that the last parameter is the return type. As @RikkiGibson says I think we need to make some tooling improvements there in general. But I do not think this proposal makes `Func<...>` worse in any way. and generic delegates are what most libraries are using because, as you say, they don't care that the function is doing they just care that the signature matches.

As soon as you can provide a description of that the implementation of a delegate type is going to be doing then I am fine with you defining your own delegate type. But in practice most people do not re-define generic delegates with custom names or have methods that have multiple arguments that are delegate types.
